File: legal-contract-pipeline/backend/app/services/vector_store.py
"""Pinecone client wrapper for RAG clause retrieval."""

from __future__ import annotations

import hashlib
from typing import Any

from pinecone import Pinecone, ServerlessSpec
from sentence_transformers import SentenceTransformer

from app.core.config import settings
from app.core.logging import get_logger

logger = get_logger(__name__)

_EMBEDDING_MODEL = "all-MiniLM-L6-v2"
_VECTOR_DIM = 384


class VectorStoreService:
    """Wrapper around Pinecone for clause embedding + retrieval."""

    def __init__(self) -> None:
        self._pc = Pinecone(api_key=settings.PINECONE_API_KEY)
        self._index = self._get_or_create_index()
        self._encoder = SentenceTransformer(_EMBEDDING_MODEL)

    def _get_or_create_index(self):
        existing = [i.name for i in self._pc.list_indexes()]
        if settings.PINECONE_INDEX not in existing:
            self._pc.create_index(
                name=settings.PINECONE_INDEX,
                dimension=_VECTOR_DIM,
                metric="cosine",
                spec=ServerlessSpec(cloud="aws", region=settings.PINECONE_ENV),
            )
            logger.info("pinecone_index_created", index=settings.PINECONE_INDEX)
        return self._pc.Index(settings.PINECONE_INDEX)

    # ------------------------------------------------------------------ #
    # Query                                                                #
    # ------------------------------------------------------------------ #

    async def query(
        self,
        text: str,
        top_k: int = 3,
        filter: dict | None = None,
    ) -> list[dict[str, Any]]:
        """Return top-k similar clauses with metadata."""
        vec = self._embed(text)
        response = self._index.query(
            vector=vec,
            top_k=top_k,
            filter=filter,
            include_metadata=True,
        )
        return [
            {
                "id": match.id,
                "score": round(match.score, 4),
                "text": match.metadata.get("text", ""),
                "type": match.metadata.get("type", ""),
                "risk_score": match.metadata.get("risk_score"),
            }
            for match in response.matches
        ]

    # ------------------------------------------------------------------ #
    # Upsert                                                               #
    # ------------------------------------------------------------------ #

    def upsert_clause(
        self,
        text: str,
        clause_type: str,
        risk_score: int,
        contract_id: str,
        clause_id: int,
    ) -> str:
        """Embed and upsert a clause with its risk score for future RAG."""
        vec = self._embed(text)
        doc_id = hashlib.sha256(f"{contract_id}:{clause_id}".encode()).hexdigest()[:32]

        self._index.upsert(
            vectors=[
                {
                    "id": doc_id,
                    "values": vec,
                    "metadata": {
                        "text": text[:1000],  # Pinecone metadata limit
                        "type": clause_type,
                        "risk_score": risk_score,
                        "contract_id": contract_id,
                    },
                }
            ]
        )
        return doc_id

    # ------------------------------------------------------------------ #
    # Embedding                                                            #
    # ------------------------------------------------------------------ #

    def _embed(self, text: str) -> list[float]:
        return self._encoder.encode(text, normalize_embeddings=True).tolist()

File: pipeline-worker/src/services/vector-store.ts
import { Pinecone } from '@pinecone-database/pinecone';
import { config } from '../config';
import crypto from 'crypto';

let pipeline: any = null;

async function getEncoder() {
  if (pipeline) return pipeline;
  try {
    // Dynamically import @xenova/transformers to keep it optional
    const { pipeline: loadPipeline } = await (import('@xenova/transformers') as any);
    pipeline = await loadPipeline('feature-extraction', 'Xenova/all-MiniLM-L6-v2');
    return pipeline;
  } catch (err) {
    console.warn('[vector_store] Failed to load @xenova/transformers, RAG embeddings will be mocked.');
    return null;
  }
}

export class VectorStoreService {
  private pc: Pinecone | null = null;
  private index: any = null;

  constructor() {
    if (config.pineconeApiKey) {
      this.pc = new Pinecone({ apiKey: config.pineconeApiKey, environment: config.pineconeEnv });
    }
  }

  private async getIndex() {
    if (this.index) return this.index;
    if (!this.pc) return null;

    try {
      const indexName = config.pineconeIndex;
      const indexes = await this.pc.listIndexes();
      const existing = (Array.isArray(indexes) ? indexes : (indexes as any).indexes || []).map((i: any) => typeof i === 'string' ? i : i.name);

      if (!existing.includes(indexName)) {
        await this.pc.createIndex({
          name: indexName,
          dimension: 384,
          metric: 'cosine',
        });
        console.log(`[vector_store] Pinecone index ${indexName} created`);
      }
      this.index = this.pc.index(indexName);
      return this.index;
    } catch (err: any) {
      console.warn(`[vector_store] Failed to initialize Pinecone index: ${err.message}`);
      return null;
    }
  }

  async query(text: string, topK: number = 3, filter?: any): Promise<any[]> {
    try {
      const index = await this.getIndex();
      if (!index) return [];

      const vec = await this.embed(text);
      if (!vec) return [];

      const queryRequest: any = {
        vector: vec,
        topK,
        includeMetadata: true,
      };
      if (filter) {
        queryRequest.filter = filter;
      }

      const response = await index.query(queryRequest);
      return (response.matches || []).map((match: any) => ({
        id: match.id,
        score: Math.round(match.score * 10000) / 10000,
        text: match.metadata?.text || '',
        type: match.metadata?.type || '',
        risk_score: match.metadata?.risk_score,
      }));
    } catch (err: any) {
      console.warn(`[vector_store] Query failed: ${err.message}`);
      return [];
    }
  }

  async upsertClause(
    text: string,
    clauseType: string,
    riskScore: number,
    contractId: string,
    clauseId: number
  ): Promise<string> {
    const docId = crypto.createHash('sha256').update(`${contractId}:${clauseId}`).digest('hex').substring(0, 32);
    try {
      const index = await this.getIndex();
      if (!index) return docId;

      const vec = await this.embed(text);
      if (!vec) return docId;

      await index.upsert([
        {
          id: docId,
          values: vec,
          metadata: {
            text: text.substring(0, 1000), // Pinecone metadata limit
            type: clauseType,
            risk_score: riskScore,
            contract_id: contractId,
          },
        },
      ]);
    } catch (err: any) {
      console.warn(`[vector_store] Upsert failed: ${err.message}`);
    }
    return docId;
  }

  private async embed(text: string): Promise<number[] | null> {
    const encoder = await getEncoder();
    if (!encoder) {
      // Return a dummy 384-dim array if encoder is unavailable
      return Array(384).fill(0).map(() => Math.random() - 0.5);
    }
    try {
      const output = await encoder(text, { pooling: 'mean', normalize: true });
      return Array.from(output.data);
    } catch (err: any) {
      console.warn(`[vector_store] Embedding generation failed: ${err.message}`);
      return null;
    }
  }
}
export const vectorStoreService = new VectorStoreService();
